File: Greedy/asteroidsDestroyed.js
const asteroidsDestroyed = (mass, asteroids) => {
  const asteroidsSorted = asteroids.toSorted((a, b) => a - b);
  for (const asteroid of asteroidsSorted) {
    if (asteroid > mass) return false;
    mass += asteroid;
  }

  return true;
};

// Time Complexity: O(n log n) due to sorting the asteroids array, 
// where n is the length of the asteroids array.
// Space Complexity: O(n) due to the space used by the sorted asteroids array, 
// could be O(1) if we sort the original array in place.

File: SlidingWindow/asteroids_destroyed.py
from typing import List


class AsteroidsDestroyed:
    def asteroidsDestroyed(self, mass: int, asteroids: List[int]) -> bool:
        asteroids.sort()
        planet = mass
        for asteroid in asteroids:
            if asteroid > planet:
                return False
            planet += asteroid

        return True

    # Time complexity: O(n log n) due to the sorting step, where n is the number of asteroids.
    # Space complexity: O(1) if we ignore the space used by the sorting algorithm,
    # otherwise O(n) due to the space used by the sorting algorithm.
